packet/src/pacer.rs

```diff
@@ -0,0 +1,131 @@
+use std::time::{Duration, Instant};
+
+pub struct LeakyBucketPacer {
+    /// Target bitrate in bits per second.
+    target_bitrate: u64,
+    /// The last time we refreshed the budget.
+    last_budget_increase: Option<Instant>,
+    /// The current budget in bytes.
+    budget: usize,
+    /// The current pacing i.e. how frequently we refresh the budget.
+    pacing: Duration,
+}
+
+pub enum Decision {
+    Emit,
+    Hold,
+}
+
+impl LeakyBucketPacer {
+    pub fn set_target_bitrate(&mut self, new_rate: u64) {
+        self.target_bitrate = new_rate;
+    }
+
+    pub fn next_interval(&self) -> Option<Instant> {
+        self.last_budget_increase.map(|l| l + self.pacing)
+    }
+
+    pub fn decide(&mut self, data: &[u8], now: Instant) -> Decision {
```

Comment:
Can this take `len: usize`, so we do't need to pass the data itself?

Reply:
I suppose so